File: jwe-implementation/java/jwe.java
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import com.nimbusds.jose.*;
import com.nimbusds.jose.crypto.*;
import com.nimbusds.jose.jwk.RSAKey;

import java.io.File;
import java.nio.file.Files;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Security;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;

public class JweEncryption {

    static {
        Security.addProvider(new BouncyCastleProvider());
    }

    public static String jweEncrypt(String publicKeyPem, String payload) throws Exception {
        PublicKey publicKey = loadPublicKey(publicKeyPem);

        // Create RSA key from the public key
        RSAKey rsaKey = new RSAKey.Builder(publicKey).build();

        // Create the JWE header and specify:
        // RSA-OAEP-256 as the encryption algorithm
        // A256GCM as the content encryption algorithm
        JWEHeader header = new JWEHeader.Builder(JWEAlgorithm.RSA_OAEP_256, EncryptionMethod.A256GCM)
                .contentType("text/plain")
                .build();

        // Create the JWE object and encrypt it
        Payload jwePayload = new Payload(payload);
        JWEObject jweObject = new JWEObject(header, jwePayload);

        RSAEncrypter encrypter = new RSAEncrypter(rsaKey);
        jweObject.encrypt(encrypter);

        // Serialize to compact form
        return jweObject.serialize();
    }

    public static String jweDecrypt(String privateKeyPem, String jweEncryptedPayload) throws Exception {
        PrivateKey privateKey = loadPrivateKey(privateKeyPem);

        // Parse the JWE string
        JWEObject jweObject = JWEObject.parse(jweEncryptedPayload);

        // Decrypt the JWE
        RSADecrypter decrypter = new RSADecrypter(privateKey);
        jweObject.decrypt(decrypter);

        // Extract the payload
        return jweObject.getPayload().toString();
    }

    private static PublicKey loadPublicKey(String pem) throws Exception {
        String publicKeyPEM = pem.replace("-----BEGIN PUBLIC KEY-----", "")
                .replace("-----END PUBLIC KEY-----", "")
                .replaceAll("\\s+", "");
        byte[] encoded = Base64.getDecoder().decode(publicKeyPEM);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        X509EncodedKeySpec keySpec = new X509EncodedKeySpec(encoded);
        return keyFactory.generatePublic(keySpec);
    }

    private static PrivateKey loadPrivateKey(String pem) throws Exception {
        String privateKeyPEM = pem.replace("-----BEGIN PRIVATE KEY-----", "")
                .replace("-----END PRIVATE KEY-----", "")
                .replaceAll("\\s+", "");
        byte[] encoded = Base64.getDecoder().decode(privateKeyPEM);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(encoded);
        return keyFactory.generatePrivate(keySpec);
    }

    public static void main(String[] args) {
        try {
            // Load public and private keys from PEM files
            String publicKeyPem = new String(Files.readAllBytes(new File("public_key.pem").toPath()));
            String privateKeyPem = new String(Files.readAllBytes(new File("private_key.pem").toPath()));

            // Sample payload to encrypt
            String requestPayload = "{\"pan\":\"\", \"client_ref_num\":\"jwe-encryption-test\"}";

            // Encrypt the payload
            String encryptedJwe = jweEncrypt(publicKeyPem, requestPayload);
            System.out.println("Encrypted JWE: " + encryptedJwe);

            // Simulate sending an encrypted request payload
            String encryptedRequestPayload = "{\"encrypted_data\":\"" + encryptedJwe + "\"}";

            // Simulate receiving a response (decrypt the payload)
            String decryptedPayload = jweDecrypt(privateKeyPem, encryptedJwe);
            System.out.println("Decrypted Payload: " + decryptedPayload);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: jwe-encrypt-decrypt/jwe.py
import json

from jwcrypto import jwk, jwe


def encrypt(public_key: str, payload: dict) -> str:
    """Encrypts a payload using a given RSA public key in JWE format."""
    # Load the RSA public key from PEM-encoded string
    public_key_jwk = jwk.JWK.from_pem(public_key.encode('utf-8'))

    # Set the JWE protected header with encryption algorithms
    protected_header = {
        "alg": "RSA-OAEP-256",  # RSA-OAEP with SHA-256 for key encryption
        "enc": "A256GCM"  # AES-GCM with 256-bit key for content encryption
    }

    # Create the JWE object and encrypt the payload
    jwe_token = jwe.JWE(
        plaintext=json.dumps(payload).encode('utf-8'),
        recipient=public_key_jwk,
        protected=protected_header
    )

    # Serialize the encrypted JWE to compact serialization format
    return jwe_token.serialize(compact=True)


def decrypt(private_key: str, jwe_encrypted_payload: str) -> str:
    """Decrypts a JWE-encrypted payload using a given RSA private key."""
    # Load the RSA private key from PEM-encoded string
    private_key_jwk = jwk.JWK.from_pem(private_key.encode('utf-8'))

    # Deserialize the encrypted JWE and decrypt it
    jwe_token = jwe.JWE()
    jwe_token.deserialize(jwe_encrypted_payload, key=private_key_jwk)

    # Return the decrypted payload as a string
    return jwe_token.payload.decode('utf-8')
